ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicationManager.java

```diff
@@ -175,9 +182,55 @@ protected ReplicaThread getReplicaThread(String threadName, FindTokenHelper find
-        leaderBasedReplicationAdmin, this.persistor, cloudDestination, properties);
+        notification, storeKeyConverter, transformer, replicatingOverSsl, datacenterName, responseHandler, time,
+        replicaSyncUpManager, skipPredicate, leaderBasedReplicationAdmin, this.persistor, cloudDestination, properties);
+  }
+
+  @Override
+  public void retrieveReplicaTokensAndPersistIfNecessary(String mountPath) {
+    // nothing to do as tokens are loaded in reloadReplicationTokenIfExists() when helix adds replica
+  }
+
+  @Override
+  public int reloadReplicationTokenIfExists(ReplicaId localReplica, List<RemoteReplicaInfo> peerReplicas)
+      throws ReplicationException {
+    AzureCloudConfig azureCloudConfig = new AzureCloudConfig(properties);
+    // For each replica of a partition
+    for (RemoteReplicaInfo replicaInfo : peerReplicas) {
+      String partitionKey = String.valueOf(replicaInfo.getReplicaId().getPartitionId().getId());
+      String rowKey = replicaInfo.getReplicaId().getDataNodeId().getHostname();
+      // First look for the token in the new place - the table
+      TableEntity row = cloudDestination.getTableEntity(azureTableNameReplicaTokens, partitionKey, rowKey);
+      if (row == null) {
+        // If token is not found on the new place, then look for it in the old place
+        // TODO: Remove this code after all nodes have migrated to using the table, tokenReloadWarnCount == 0
+        vcrMetrics.tokenReloadWarnCount.inc();
+        if (super.reloadReplicationTokenIfExists(localReplica, peerReplicas) > 0) {
+          // If token is not found in the old place too, then that's a real error !
+          azureMetrics.absTokenRetrieveFailureCount.inc();
+          logger.error("Failed to retrieve tokens for peer replicas of local replica {}", localReplica);
+        }
+        return (int) azureMetrics.absTokenRetrieveFailureCount.getCount();
+      }
+      FindTokenFactory findTokenFactory =
+          tokenHelper.getFindTokenFactoryFromReplicaType(ReplicaType.DISK_BACKED);
+      try {
+        DataInputStream inputStream = new DataInputStream(
+            new ByteArrayInputStream((byte[]) row.getProperty(BINARY_TOKEN)));
+        replicaInfo.setToken(findTokenFactory.getFindToken(inputStream));
+        String replUntil = (String) row.getProperty(REPLICATED_UNITL_UTC);
+        long time = replUntil.equals(String.valueOf(Utils.Infinite_Time)) ? -1 : DATE_FORMAT.parse(replUntil).getTime();
+        replicaInfo.setReplicatedUntilUTC(time);
+      } catch (Throwable t) {
+        // log and metric
+        azureMetrics.absTokenRetrieveFailureCount.inc();
+        logger.error("Failed to deserialize token for peer replica {} due to {}", replicaInfo, t.toString());
+        t.printStackTrace();
+        replicaInfo.setToken(findTokenFactory.getNewFindToken());
+        replicaInfo.setReplicatedUntilUTC(-1);
+      }
+    }
+    return (int) azureMetrics.absTokenRetrieveFailureCount.getCount();
```

Comment:
nit: Do we need return type to be integer or `boolean` is okay? Seems like old code also needs to be updated.

Reply:
This is just to keep in line with old code, the return value isn't really used except for emitting it as another redundant metric.